drivers/sensor/htu21d/htu21d.c

```diff
@@ -0,0 +1,310 @@
+	uint16_t tmp;
+	uint8_t crc;
+	int ret;
+
+	ret = htu21d_write(dev, HTU21D_HUMIDITY_MEASUREMENT_NO_HOLD_MASTER);
+	if (ret < 0) {
+		LOG_DBG("Humidity measurement failed: %d", ret);
+		return ret;
+	}
+
+	/* Wait for the measure to be ready */
+	k_sleep(K_MSEC(16)); /* Max at 12 bit resolution */
+
+	ret = htu21d_read(dev, buf, sizeof(buf));
+	if (ret < 0) {
+		LOG_DBG("Read failed: %d", ret);
+		return ret;
+	}
+
+	/* Check for the status */
+	ret = buf[1] & 0x03;
+	if (ret == 0x11) {
+		LOG_DBG("Status failed: %d", ret);
+		return -1;
+	}
+
+	tmp = sys_get_be16(&buf[0]);
+	crc = htu21d_compute_crc(tmp);
+	if (crc != buf[2]) {
+		LOG_DBG("CRC invalid: %02x, expected: %02x", crc, buf[2]);
+		return -1;
+	}
+
+	return tmp;
+}
+
+static int htu21d_temperature_fetch(const struct device *dev)
+{
+	struct htu21d_data *data = dev->data;
+	uint8_t buf[3] = { -1 };
+	uint16_t tmp;
+	uint8_t crc;
+	int ret;
+
+	ret = htu21d_write(dev, HTU21D_TEMPERATURE_MEASUREMENT_NO_HOLD_MASTER);
+	if (ret < 0) {
+		LOG_DBG("Temperature measurement failed: %d", ret);
+		return ret;
+	}
+
+	/* Wait for the measure to be ready */
+	k_sleep(K_MSEC(50)); /* Max at 14 bit resolution */
+
+	ret = htu21d_read(dev, buf, sizeof(buf));
+	if (ret < 0) {
+		LOG_DBG("Read failed: %d", ret);
+		return ret;
+	}
+
+	/* Check for the status */
+	ret = buf[1] & 0x03;
+	if (ret == 0x11) {
+		LOG_DBG("Status failed: %d", ret);
+		return -1;
+	}
+
+	tmp = sys_get_be16(&buf[0]);
+	crc = htu21d_compute_crc(tmp);
+	if (crc != buf[2]) {
+		LOG_DBG("CRC invalid: %02x, expected: %02x", crc, buf[2]);
+		return -1;
+	}
+
+	return tmp;
+}
+
+static int htu21d_sample_fetch(const struct device *dev,
+			       enum sensor_channel chan)
+{
+	struct htu21d_data *data = dev->data;
+	uint8_t buf[3];
+	uint16_t tmp;
+	int ret;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+
+#ifdef CONFIG_PM_DEVICE
+	enum pm_device_state state;
+	(void)pm_device_state_get(dev, &state);
+	/* Do not allow sample fetching from suspended state */
+	if (state == PM_DEVICE_STATE_SUSPENDED)
+		return -EIO;
+#endif
+
+	ret = htu21d_humidity_fetch(dev);
+	if (ret < 0)
+		return ret;
+	data->humidity_raw_val = ret;
+
+	ret = htu21d_temperature_fetch(dev);
+	if (ret < 0)
+		return ret;
+	data->temperature_raw_val = ret;
+
+	return 0;
+}
+
+static int htu21d_channel_get(const struct device *dev,
+			      enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	struct htu21d_data *data = dev->data;
+	double tmp;
+	int ret;
+
+	switch (chan) {
+	case SENSOR_CHAN_HUMIDITY:
+		/*
+		 * The documentation says the following about the conversion at
+		 * section Conversion of signal output p.15:
+		 *
+		 * Relative Humidity conversion
+		 *
+		 * With the relative humidity signal output SRH, the relative
+		 * humidity is obtained by the following formula (result in
+		 * %RH), no matter which resolution is chosen:
+		 *
+		 * RH = -6 + 125 * SRH / 2^16
+		 */
+		tmp = ((125.0 * data->humidity_raw_val) / 65536.0) - 6.0;
```

Comment:
You don't need floating point math for this

Reply:
do you prefer the former solution with `long long`?

``` c
static int htu21d_channel_get(const struct device *dev,
			      enum sensor_channel chan,
			      struct sensor_value *val)
{
	struct htu21d_data *data = dev->data;
	long long tmp;

	switch (chan) {
	case SENSOR_CHAN_HUMIDITY:
		tmp = data->humidity_raw_val;
		tmp = ((125000000LL * tmp) / 65536LL) - 6000000LL;
		val->val1 = tmp / 1000000;
		val->val2 = tmp % 1000000;
		break;
	case SENSOR_CHAN_AMBIENT_TEMP:
		tmp = data->temperature_raw_val;
		tmp = ((175720000LL * tmp) / 65536LL) - 46850000LL;
		val->val1 = tmp / 1000000;
		val->val2 = tmp % 1000000;
		break;
	default:
		return -EINVAL;
	}

	return 0;
}
```

Reply:
No need to use 64-bit math.  64-bit integer divide is really slow.

The original value is only at most 14 bits for temp or 12 for RH, multiplying by 175.72 (temp) or 125 (RH) can not overflow 32 bits.

In the Si7006 driver, the code is quite silly: it rounds to an integer, then multiplies by 1000000, then divides by 1000000.  Obviously, those last two steps just cancel out.

This is better.   But nonetheless, it is still possible to keep the fractional value without 64-bit math.  The key is separate the fractional and integer components *before* converting into micro-degrees or micro-percent.

```c
const uint32_t rh_x16 = data->humidity_raw_val * 125U; // This is RH * 2¹⁶
const int16_t rh_int = rh_x16 >> 16;        // Upper 16 bits are the integer component
const uint16_t rh_frac = rh_x16 & 0xffffU;  // Lower 16 bits are fractional part
val->val1 = rh_int - 6;
// val->val2 = rh_frac * 1000000 / 65536;   // This could overflow
val->val2 = rh_frac * 15625 / 1024;         // But if we divide both values by 64, it doesn't!

//  The split 2s-compliment/BCD format is really annoying to work with!
if (val->val1 < 0) {
  val->val2 = -val->val2;
}
```

This avoids all 64-bit math and all integer division.  Only shifts are needed.